src/common/cockpitframe.c

```diff
@@ -139,103 +144,155 @@ cockpit_frame_write (int fd,
   return ret;
 }
 
-static void *
-xrealloc (void *old,
-          size_t length)
+/* read_exactly:
+ * @fd: a blocking file descriptor
+ * @buffer: where to read to
+ * @required_size: the exact number of bytes to read
+ * @was_eof: if EOF was encountered
+ *
+ * Reads exactly @required_size bytes from @fd into @buffer.  If the
+ * correct number of bytes are read, then %TRUE is returned.
+ *
+ * On failure, returns %FALSE.  If the failure was due to an underlying
+ * read error, then this error will be stored into errno.  If the
+ * failure was due to an incorrect number of bytes being read, then
+ * errno will be set to %EBADMSG.
+ *
+ * The only permitted exception is if @was_eof is non-%NULL.  In that
+ * case, if exactly 0 bytes are read from @fd (ie: EOF at the start of
+ * the message) then @was_eof will be set to %TRUE and the call will
+ * return successfully.  Otherwise, @was_eof will be set to %FALSE on
+ * successful completion.
+ *
+ * If the required number of bytes couldn't be read, then any number of
+ * bytes may have been read from @fd.  The only reasonable thing to do
+ * at this point is to treat the connection as broken, and close @fd.
+ * In particular, this is why @fd should never be non-blocking.
+ */
+static bool
+read_exactly (int fd,
+              unsigned char *buffer,
+              size_t required_size,
+              bool *was_eof)
```

Comment:
Nice! It feels like this should be in a very central util library, this seems generally useful. (No need to move it around now, of course)

Reply:
There's a "write all" type function in `cockpitframe.c` as well, and it gets exported via the header.  I considered doing the same to this (with a nicer name) but will wait until we have an actual user elsewhere.

Note there's some potential commonality/overlap here with the `socket-io.c` functions in `cockpit-tls`...